File: 2019/Round_A/Training/training.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	const maxCapacity = 1024 * 1024

	var (
		t     int
		input []string
	)

	// Read T (Number of test cases)
	fmt.Scanln(&t)

	scanner := bufio.NewScanner(os.Stdin)
	buf := make([]byte, maxCapacity)
	scanner.Buffer(buf, maxCapacity)
	for i := 0; i < 2*t; i++ {
		scanner.Scan()
		input = append(input, scanner.Text())
	}

	// Deal with each test case
	for i := 0; i < t; i++ {
		// Read N (Number of students) & P (Number of players)
		var n, p int
		fmt.Sscan(input[2*i], &n, &p)

		s := strings.Split(input[2*i+1], " ")
		skills := []int{}

		// Read skill of each student
		for j := 0; j < n; j++ {
			skill, _ := strconv.Atoi(s[j])
			skills = append(skills, skill)
		}

		// Sort skill in descending order
		sort.Ints(skills)

		skillSums := []int{0}

		// Calculate accumulative sum
		for j := 0; j < n; j++ {
			skillSums = append(skillSums, skillSums[len(skillSums)-1]+skills[n-j-1])
		}

		// Set original value for the answer
		minHours := -1
		for j := 0; j < n-p+1; j++ {
			// Calculate current skill
			currentSkillSum := skillSums[j+p] - skillSums[j]

			// Calculate target skill
			targetSkillSum := skills[n-j-1] * p

			// Calculate hours required
			hours := targetSkillSum - currentSkillSum

			// Update answer with new minimal
			if hours < minHours || minHours < 0 {
				minHours = hours
			}
		}

		// Print the answer
		fmt.Printf("Case #%d: %d\n", i+1, minHours)
	}
}

File: 2019/Round_A/Training/training.js
const readline = require('readline')
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false,
})

let t = 0
let n = 0
let p = 0
let s = []
let tIndex = 0

const main = () => {
  rl.on('line', function(line) {
    if (t === 0) {
      t = Number(line)
      return
    }

    if (n === 0) {
      ;[n, p] = line.split(' ').map(char => Number(char))
      return
    }

    if (s.length === 0) {
      const sRaw = line.split(' ')
      for (let i = 0; i < sRaw.length; i++) {
        s.push(Number(sRaw[i]))
      }

      s.sort((a, b) => b - a)

      const accuSum = [0]
      for (let i = 0; i < s.length; i++) {
        accuSum.push(accuSum[accuSum.length - 1] + s[i])
      }

      let minHours = -1
      for (let i = 0; i < n - p + 1; i++) {
        currentSkill = accuSum[i + p] - accuSum[i]

        targetSkill = p * s[i]

        hours = targetSkill - currentSkill

        if (hours < minHours || minHours < 0) {
          minHours = hours
          continue
        }
      }

      tIndex++
      console.log(`Case #${tIndex}: ${minHours}`)
      n = 0
      p = 0
      s = []
    }

    if (tIndex === t) {
      rl.close()
      return
    }
  })
}

main()
